### PR TITLE
Added matrix scan for android

### PR DESCRIPTION
**Pull request by robertnawrot**
Added matrix scan functionality to the android version of React Native plugin

### CODE REVIEW

File: android/src/main/java/com/scandit/reactnative/BarcodePicker.kt [hunk 1]
```diff
@@ -81,18 +93,55 @@ class BarcodePicker : SimpleViewManager<BarcodePicker>(), OnScanListener, TextRe
...
+                "onRecognizeNewCodes", (newlyTrackedCodesToMap(newlyTrackedCodes)))
+        didProcessLatch.await()
+        for (id in idsToReject) {
+            scanSession.rejectTrackedCode(scanSession.trackedCodes[id.toLong()])
+        }
+        idsToReject.clear()
+    }
+
     override fun didScan(scanSession: ScanSession?) {
+        if (isMatrixScanEnabled || scanSession ==  null) {
```
[the remark this answers is not in the document]
Comment (robertnawrot) [reply android/src/main/java/com/scandit/reactnative/BarcodePicker.kt]: I know, but you have to do it from the frame process callback or you'll get an exception.

File: android/src/main/java/com/scandit/reactnative/BarcodePicker.kt [hunk 2]
```diff
@@ -81,18 +93,55 @@ class BarcodePicker : SimpleViewManager<BarcodePicker>(), OnScanListener, TextRe
...
+        val context = picker?.context as ReactContext?
+
+        val trackedCodes = scanSession.trackedCodes
+        val newlyTrackedCodes = ArrayList<TrackedBarcode>()
+        val recognizedCodeIds = HashSet<Long>()
+
+        for (entry in trackedCodes.entries) {
+            if (entry.value.isRecognized) {
+                recognizedCodeIds.add(entry.key)
+                if (lastFrameRecognizedIds.add(entry.key)) {
```
Comment (lucatorella) [inline android/src/main/java/com/scandit/reactnative/BarcodePicker.kt]: why not a simple contains here? 😊 `!lastFrameRecognizedIds.contains(entry.key)`

File: android/src/main/java/com/scandit/reactnative/BridgeHelpers.kt [hunk 3]
```diff
@@ -147,4 +177,10 @@ fun ReadableMap.toJson(): JSONObject {
     return jsonObject
 }
 
-fun settingsFromMap(map: ReadableMap): ScanSettings = ScanSettings.createWithJson(map.toJson())
+fun settingsFromMap(map: ReadableMap): ScanSettings {
+    val settings = ScanSettings.createWithJson(map.toJson())
+    if (settings.isMatrixScanEnabled) {
+        settings.codeDuplicateFilter = 0
```
Comment (lucatorella) [inline android/src/main/java/com/scandit/reactnative/BridgeHelpers.kt]: @robertnawrot this is not necessary, can you remove it?

Comment (robertnawrot) [reply android/src/main/java/com/scandit/reactnative/BridgeHelpers.kt]: We investigated it with @moritz-scandit and @biasmv and it seemed necessary. At least for React Native plugin.

Comment (moritz-scandit) [reply android/src/main/java/com/scandit/reactnative/BridgeHelpers.kt]: Actually I don't think it's necessary, the sdk already sets it to this by itself, which is why we got all the didScan events in the first place. So you should be able to remove this.

File: android/src/main/java/com/scandit/reactnative/BarcodePicker.kt [hunk 4]
```diff
@@ -120,12 +171,28 @@ class BarcodePicker : SimpleViewManager<BarcodePicker>(), OnScanListener, TextRe
...
-        latch = CountDownLatch(1)
+        didScanLatch.countDown()
+        didScanLatch = CountDownLatch(1)
+    }
+
+    private fun finishDidProcessCallback(args: ReadableArray?) {
+        if (args?.getBoolean(0) == true)
+            picker?.stopScanning()
+        if (args?.getBoolean(1) == true)
+            picker?.pauseScanning()
```
[the remark this answers is not in the document]
Comment (robertnawrot) [reply android/src/main/java/com/scandit/reactnative/BarcodePicker.kt]: That's true, I remember we discussed this. I'll try to include all those fixes before the release branch-off.

**Pull request merged**